### PR TITLE
Layered docker images and improved docker-compose deployment

### PR DESCRIPTION
**Pull request by Zeerooth**
Changes include:
- multi-layered build process for docker:
  - use cargo chef to gather dependencies into a recipe
  - use cargo chef to install said dependencies
  - build kitsune
  - build frontend
  - include the app and frontend files in the final image
This makes it so only the project libraries and binaries are rebuilt after making changes, instead of the whole dependency tree speeding up the build process from around 15m to 3m and saving space.
- docker-compose can now receive the app config as a mounted readonly file, like so
`KITSUNE_CONFIG=/home/user/kitsune/config.docker.toml docker-compose up`
- I made kitsune depend on db and redis in the docker compose stack because otherwise it would sometimes panic if these components haven't properly initialized before the app started

### CODE REVIEW

**Comment by aumetra** [comment]
Another thing that might be worth considering, we talked a bit about including the Docker image building into our Nix flake instead of keeping it as separate flows.

Essentially just unifying all of our building into Nix.

**Comment by aumetra** [comment]
(Test failure is unrelated, one of the test dependent on an external service broke (oh, who could have predicted this?). I will just force merge once we got everything figured out)

**Review by aumetra — CHANGES_REQUESTED** [review]

File: kitsune/Dockerfile.dev [hunk 1]
```diff
@@ -1,11 +1,27 @@
-FROM rust:1-alpine AS build
+FROM rust:1-alpine AS base
 RUN rustup default nightly
```
[the remark this answers is not in the document]
Comment (Zeerooth) [reply kitsune/Dockerfile.dev]: That's true, I haven't noticed that kitsune uses stable now. I fixed it, thanks!

**Comment by aumetra** [comment]
I thought we maybe could utilize the Nix caching instead. 

`cargo2nix` could provide us with the caching, and then we just build a simple image taking the flake outputs we defined for the frontend and backend as package inputs.

So the Docker functions will just take the binaries out of the Nix store.

Something like this (of course with a `CMD` defined):

```nix
pkgs.dockerTools.buildLayeredImage {
    name = "kitsune";
    tag = "latest";
    packages = [ backend ];
}
```

(where the `backend` reference references the output of the flake)

**Comment by Zeerooth** [comment]
Is `packages` an actual argument for `buildLayeredImage`? When I try to use it I get 
```
error: 'streamLayeredImage' at /nix/store/aw6kmwd8a02n2c1wysrfk2q31brlmqdz-source/pkgs/build-support/docker/default.nix:841:5 called with unexpected argument 'packages'
```
and I can't find anything in the docs.

I tried to build an image with the following flake:
```
{
  inputs = {
    flake-utils.url = "github:numtide/flake-utils";
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    rust-overlay.url = "github:oxalica/rust-overlay";
  };
  outputs = { self, flake-utils, nixpkgs, rust-overlay }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        overlays = [ (import rust-overlay) ];
        pkgs = import nixpkgs {
          inherit overlays system;
        };
        rustPlatform = pkgs.makeRustPlatform {
          cargo = pkgs.rust-bin.stable.latest.minimal;
          rustc = pkgs.rust-bin.stable.latest.minimal;
        };
        baseDependencies = with pkgs; [
          openssl
          pkg-config
          protobuf
          sqlite
          zlib
        ];
        cargoConfig = builtins.fromTOML (builtins.readFile ./.cargo/config.toml);  # TODO: Set the target CPU conditionally
        cargoToml = builtins.fromTOML (builtins.readFile ./Cargo.toml);
        src = pkgs.lib.cleanSourceWith {
          src = pkgs.lib.cleanSource ./.;
          filter = name: type:
            let baseName = baseNameOf (toString name);
            in !(baseName == "flake.lock" || pkgs.lib.hasSuffix ".nix" baseName);
        };
        version = cargoToml.workspace.package.version;
        basePackage = {
          inherit version src;

          meta = {
            description = "ActivityPub-federated microblogging";
            homepage = "https://joinkitsune.org";
          };

          cargoLock = {
            lockFile = ./Cargo.lock;
            allowBuiltinFetchGit = true;
          };

          nativeBuildInputs = baseDependencies;

          PKG_CONFIG_PATH = "${pkgs.openssl.dev}/lib/pkgconfig"; # Not sure why this is broken but it is
          RUSTFLAGS = builtins.concatStringsSep " " cargoConfig.build.rustflags; # Oh god help.

          checkFlags = [
            # Depend on creating an HTTP client and that reads from the systems truststore
            # Because nix is fully isolated, these types of tests fail
            #
            # Some (most?) of these also depend on the network? Not good??
            "--skip=activitypub::fetcher::test::federation_allow"
            "--skip=activitypub::fetcher::test::federation_deny"
            "--skip=activitypub::fetcher::test::fetch_actor"
            "--skip=activitypub::fetcher::test::fetch_note"
            "--skip=resolve::post::test::parse_mentions"
            "--skip=webfinger::test::fetch_qarnax_ap_id"
            "--skip=basic_request"
            "--skip=json_request"
          ];
        };
        dockerImage = pkgs.dockerTools.buildLayeredImage {
          name = "kitsune";
          tag = "latest";
          config.Cmd = [ "kitsune" ];
        };
      in
      {
        packages = rec {
          default = main;
          cli = rustPlatform.buildRustPackage (basePackage // {
            pname = "kitsune-cli";
            cargoBuildFlags = "-p kitsune-cli";
          });
          main = rustPlatform.buildRustPackage (basePackage // {
            pname = "kitsune";
            buildFeatures = [ "meilisearch" ];
            cargoBuildFlags = "-p kitsune";
          });
          search = rustPlatform.buildRustPackage (basePackage // {
            pname = "kitsune-search";
            cargoBuildFlags = "-p kitsune-search";
          });
          frontend = pkgs.mkYarnPackage {
            inherit version;

            src = "${src}/kitsune-fe";

            buildPhase = ''
              yarn --offline build
            '';

            installPhase = ''
              mkdir -p $out
              cp -R deps/kitsune-fe/dist $out
            '';

            distPhase = "true";
          };
          docker = dockerImage;
        };
        devShells = rec {
          default = backend;
          backend = pkgs.mkShell {
            buildInputs = with pkgs; [
              cargo-insta
              dhall
              diesel-cli
              redis
              rust-bin.stable.latest.default
            ]
            ++
            baseDependencies;
          };
          frontend = pkgs.mkShell {
            buildInputs = with pkgs; [
              nodejs
              yarn
            ];
          };
        };
      }
    ) // {
      overlays = rec {
        default = kitsune;
        kitsune = (import ./overlay.nix self);
      };

      nixosModules = rec {
        default = kitsune;
        kitsune = (import ./module.nix);
      };
    };
}
```
build and load it with `docker load < result` but it just results in an empty image which I don't know how to correct.

Could we just do a merge with my changes for now, make an issue for moving the docker build into a nix flake and possibly solve it in another PR please?

**Comment by aumetra** [comment]
Sure, of course. The Nix Docker stuff was just me putting out some ideas on here.

Thank you for fixing up the Docker/Docker Compose things

**Pull request merged**